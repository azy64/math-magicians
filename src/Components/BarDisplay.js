import React from 'react';
import PropTypes from 'prop-types';

export default class BarDisplay extends React.Component {
  constructor(props) {
    super(props);
    this.state = {};
  }

  render = () => {
    const { content } = this.props;
    return (
      <div className="bar-display">
        <div className="writting" role="none">
          {content}
        </div>
      </div>
    );
  }
}
BarDisplay.propTypes = {
  content: PropTypes.string.isRequired,
};
