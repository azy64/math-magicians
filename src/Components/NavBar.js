import React from 'react';
import { NavLink } from 'react-router-dom';

export default function NavBar() {
  return (
    <div className="navbar">
      <div className="title">
        <h1>Math Magicians</h1>
      </div>
      <div className="link">
        <ul>
          <li>
            <NavLink to="/" exact>Home</NavLink>
          </li>
          <li>
            <NavLink to="/calc" exact>Calculator</NavLink>
          </li>
          <li>
            <NavLink to="/quote" exact>Quote</NavLink>
          </li>
        </ul>
      </div>
    </div>
  );
}
