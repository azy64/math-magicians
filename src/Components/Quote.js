import React from 'react';

export default function Quote() {
  return (
    <div style={{
      width: '70%', margin: 'auto', textAlign: 'center', fontSize: 19, marginTop: '5%',
    }}
    >
      <p>
        Mathematics is about numbers, equations, computations
        <br />
        or algorithms: it is about understanding. -William Paul Thurston
      </p>
    </div>
  );
}
