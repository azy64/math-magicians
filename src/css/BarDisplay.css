.bar-display {
  width: 100%;
  height: 35px;
  background-color: #858693;
}

.bar-display .writting {
  width: 98%;
  margin: auto;
  color: white;
  text-align: end;
  line-height: 35px;
  padding-right: 10px;
}
