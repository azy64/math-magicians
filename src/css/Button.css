
.button {
  display: inline-block;
  text-align: center;
  line-height: 35px;
  border: 1px solid gray;
  cursor: pointer;
}
