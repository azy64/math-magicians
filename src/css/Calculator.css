.section-calc {
  display: grid;
  grid-template-columns: 40% 50%;
}

.calculator {
  width: 300px;
  min-height: 100px;
  margin: auto;
}
