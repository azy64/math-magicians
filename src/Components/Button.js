import React from 'react';
import PropTypes from 'prop-types';
import '../css/Button.css';

class Button extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
    };
  }

  clickHandler = (e) => {
    const txt = e.target.firstChild.textContent;
    const { content, calculate, setCalculate } = this.props;
    content(txt, calculate, setCalculate);
  }

  parseStringOp = (str) => {
    const op = [];
    const str1 = [...str];
    // we get the opration in order and get numbers
    for (let i = 0; i < str1.length; i += 1) {
      if (str1[i] === '+') op.push('+');
      if (str1[i] === '-') op.push('-');
      if (str1[i] === 'รท') op.push('รท');
      if (str1[i] === 'x') op.push('x');
      if (str1[i] === '%') op.push('%');
    }

    const reg = /[+\-รทx%]/ig;
    let tmp = str.replace(reg, ' ');
    tmp = tmp.split(' ');
    return tmp;
  }

  render() {
    const { width, text, color } = this.props;
    const styles = { backgroundColor: `${color}`, width: `${width}` };
    return (
      <div style={styles} className="button" onClick={this.clickHandler} role="button" tabIndex="0" onKeyPress={this.clickHandler}>
        <span className="text">{text}</span>
      </div>
    );
  }
}

Button.propTypes = {
  width: PropTypes.string.isRequired,
  text: PropTypes.string.isRequired,
  color: PropTypes.string.isRequired,
  content: PropTypes.func.isRequired,
  // eslint-disable-next-line react/forbid-prop-types
  calculate: PropTypes.object.isRequired,
  setCalculate: PropTypes.func.isRequired,
};

export default Button;
