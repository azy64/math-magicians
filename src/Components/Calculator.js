import React, { useState } from 'react';
import BarDisplay from './BarDisplay';
import Button from './Button';
import '../css/Calculator.css';
import calculate from '../logic/calculate';

const clickHandler = async (txt, compute, setCompute) => {
  setCompute(calculate(compute, txt));
};

const addButton = (compute, setCompute) => {
  const tab = ['AC', '+/-', '%', 'รท', '7', '8', '9', 'x', '4', '5', '6', '-', '1', '2', '3', '+', '0', '.', '='];
  const tmp = [];
  let i = 0;
  tab.forEach((signe) => {
    if (signe === '+' || signe === '=' || signe === '-' || signe === 'รท' || signe === 'x') tmp.push(<Button key={i} calculate={compute} setCalculate={setCompute} content={clickHandler} width="73px" color="#F4924B" text={signe} />);
    else if (signe === '0') tmp.push(<Button calculate={compute} setCalculate={setCompute} key={i} content={clickHandler} width="148px" color="#E0E0E0" text={signe} />);
    else tmp.push(<Button key={i} calculate={compute} setCalculate={setCompute} content={clickHandler} width="73px" color="#E0E0E0" text={signe} />);
    i += 1;
  });
  return tmp;
};
const Calculator = () => {
  const init = {
    total: null,
    next: null,
    operation: null,
  };
  const [compute, setCompute] = useState(init);
  const { next, total } = compute;
  const tmp = next || total || 0;
  return (
    <div className="calculator">
      <BarDisplay content={`${tmp}`} />
      {addButton(compute, setCompute)}
    </div>
  );
};

export default Calculator;
