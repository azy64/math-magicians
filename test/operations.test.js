import React from 'react';
import renderer from 'react-test-renderer';
import { BrowserRouter as Router } from 'react-router-dom';
import { render, screen, fireEvent } from '@testing-library/react';
import operate from '../src/logic/operate';
import calculate from '../src/logic/calculate';
import Home from '../src/Components/Home';
import NavBar from '../src/Components/NavBar';
import Button from '../src/Components/Button';
import BarDisplay from '../src/Components/BarDisplay';
import Quote from '../src/Components/Quote';
import Calculator from '../src/Components/Calculator';

describe('testing operate and calculate function', () => {
  test('testing  for string number', () => {
    expect(operate('3', '6', '+')).toBe('9');
  });
  test('testing  for numbers', () => {
    expect(operate(3, 6, '+')).toBe('9');
  });
  test('testing an unknown operation goes as expected', () => {
    expect(() => operate('4', '6', '}')).toThrow();
    expect(() => operate('4', '6', '[')).toThrow(Error);
  });
  test('testing calculate function', () => {
    const ob = { next: '6', operation: '+', total: null };
    expect(calculate(ob, '6')).toEqual({ next: '66', operation: '+', total: null });
  });
  test('testing calculate function when doing operation =', () => {
    const ob = { next: '6', operation: '+', total: '5' };
    expect(calculate(ob, '=')).toEqual({ next: null, operation: null, total: '11' });
  });
  test('testing calculate function when doing AC operation =', () => {
    const ob = { next: '6', operation: '+', total: '5' };
    expect(calculate(ob, 'AC')).toEqual({ next: null, operation: null, total: null });
  });
  test('testing calculate function when doing +/- operation =', () => {
    const ob = { next: '6', operation: '+', total: '5' };
    expect(calculate(ob, '+/-')).toEqual({ next: '-6', operation: '+', total: '5' });
  });
  it('renders correctly for Home component', () => {
    const tree = renderer
      .create(<Home />)
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
  it('renders correctly for Button component', () => {
    const tree = renderer
      .create(<Button
        width="100px"
        text="1"
        color="blue"
        content={(x) => x}
        calculate={{ total: null, next: null }}
        setCalculate={(x) => x}
      />)
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('renders correctly for BarDisplay component', () => {
    const tree = renderer
      .create(<BarDisplay content="12" />)
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
  it('renders correctly for Quote component', () => {
    const tree = renderer
      .create(<Quote />)
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
  it('renders correctly for NavBar component', () => {
    const tree = renderer
      .create(<Router><NavBar /></Router>)
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
  it('renders correctly Calculator component', () => {
    const tree = renderer.create(<Calculator />).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('should return the sum of two number form UI', () => {
    render(<Calculator />);
    fireEvent.click(screen.getByText('3'));
    fireEvent.click(screen.getByText('x'));
    fireEvent.click(screen.getByText('7'));
    fireEvent.click(screen.getByText('='));
    const result = screen.getByRole('none');
    expect(result.innerHTML).toBe('21');
  });
});
