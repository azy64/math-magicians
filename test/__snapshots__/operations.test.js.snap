// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`testing operate and calculate function renders correctly Calculator component 1`] = `
<div
  className="section-calc"
>
  <div>
    <h1>
      Let &lsquo Do some math !
    </h1>
  </div>
  <div
    className="calculator"
  >
    <div
      className="bar-display"
    >
      <div
        className="writting"
        role="none"
      >
        0
      </div>
    </div>
    <div
      className="button"
      onClick={[Function]}
      onKeyPress={[Function]}
      role="button"
      style={
        Object {
          "backgroundColor": "#E0E0E0",
          "width": "73px",
        }
      }
      tabIndex="0"
    >
      <span
        className="text"
      >
        AC
      </span>
    </div>
    <div
      className="button"
      onClick={[Function]}
      onKeyPress={[Function]}
      role="button"
      style={
        Object {
          "backgroundColor": "#E0E0E0",
          "width": "73px",
        }
      }
      tabIndex="0"
    >
      <span
        className="text"
      >
        +/-
      </span>
    </div>
    <div
      className="button"
      onClick={[Function]}
      onKeyPress={[Function]}
      role="button"
      style={
        Object {
          "backgroundColor": "#E0E0E0",
          "width": "73px",
        }
      }
      tabIndex="0"
    >
      <span
        className="text"
      >
        %
      </span>
    </div>
    <div
      className="button"
      onClick={[Function]}
      onKeyPress={[Function]}
      role="button"
      style={
        Object {
          "backgroundColor": "#F4924B",
          "width": "73px",
        }
      }
      tabIndex="0"
    >
      <span
        className="text"
      >
        รท
      </span>
    </div>
    <div
      className="button"
      onClick={[Function]}
      onKeyPress={[Function]}
      role="button"
      style={
        Object {
          "backgroundColor": "#E0E0E0",
          "width": "73px",
        }
      }
      tabIndex="0"
    >
      <span
        className="text"
      >
        7
      </span>
    </div>
    <div
      className="button"
      onClick={[Function]}
      onKeyPress={[Function]}
      role="button"
      style={
        Object {
          "backgroundColor": "#E0E0E0",
          "width": "73px",
        }
      }
      tabIndex="0"
    >
      <span
        className="text"
      >
        8
      </span>
    </div>
    <div
      className="button"
      onClick={[Function]}
      onKeyPress={[Function]}
      role="button"
      style={
        Object {
          "backgroundColor": "#E0E0E0",
          "width": "73px",
        }
      }
      tabIndex="0"
    >
      <span
        className="text"
      >
        9
      </span>
    </div>
    <div
      className="button"
      onClick={[Function]}
      onKeyPress={[Function]}
      role="button"
      style={
        Object {
          "backgroundColor": "#F4924B",
          "width": "73px",
        }
      }
      tabIndex="0"
    >
      <span
        className="text"
      >
        x
      </span>
    </div>
    <div
      className="button"
      onClick={[Function]}
      onKeyPress={[Function]}
      role="button"
      style={
        Object {
          "backgroundColor": "#E0E0E0",
          "width": "73px",
        }
      }
      tabIndex="0"
    >
      <span
        className="text"
      >
        4
      </span>
    </div>
    <div
      className="button"
      onClick={[Function]}
      onKeyPress={[Function]}
      role="button"
      style={
        Object {
          "backgroundColor": "#E0E0E0",
          "width": "73px",
        }
      }
      tabIndex="0"
    >
      <span
        className="text"
      >
        5
      </span>
    </div>
    <div
      className="button"
      onClick={[Function]}
      onKeyPress={[Function]}
      role="button"
      style={
        Object {
          "backgroundColor": "#E0E0E0",
          "width": "73px",
        }
      }
      tabIndex="0"
    >
      <span
        className="text"
      >
        6
      </span>
    </div>
    <div
      className="button"
      onClick={[Function]}
      onKeyPress={[Function]}
      role="button"
      style={
        Object {
          "backgroundColor": "#F4924B",
          "width": "73px",
        }
      }
      tabIndex="0"
    >
      <span
        className="text"
      >
        -
      </span>
    </div>
    <div
      className="button"
      onClick={[Function]}
      onKeyPress={[Function]}
      role="button"
      style={
        Object {
          "backgroundColor": "#E0E0E0",
          "width": "73px",
        }
      }
      tabIndex="0"
    >
      <span
        className="text"
      >
        1
      </span>
    </div>
    <div
      className="button"
      onClick={[Function]}
      onKeyPress={[Function]}
      role="button"
      style={
        Object {
          "backgroundColor": "#E0E0E0",
          "width": "73px",
        }
      }
      tabIndex="0"
    >
      <span
        className="text"
      >
        2
      </span>
    </div>
    <div
      className="button"
      onClick={[Function]}
      onKeyPress={[Function]}
      role="button"
      style={
        Object {
          "backgroundColor": "#E0E0E0",
          "width": "73px",
        }
      }
      tabIndex="0"
    >
      <span
        className="text"
      >
        3
      </span>
    </div>
    <div
      className="button"
      onClick={[Function]}
      onKeyPress={[Function]}
      role="button"
      style={
        Object {
          "backgroundColor": "#F4924B",
          "width": "73px",
        }
      }
      tabIndex="0"
    >
      <span
        className="text"
      >
        +
      </span>
    </div>
    <div
      className="button"
      onClick={[Function]}
      onKeyPress={[Function]}
      role="button"
      style={
        Object {
          "backgroundColor": "#E0E0E0",
          "width": "148px",
        }
      }
      tabIndex="0"
    >
      <span
        className="text"
      >
        0
      </span>
    </div>
    <div
      className="button"
      onClick={[Function]}
      onKeyPress={[Function]}
      role="button"
      style={
        Object {
          "backgroundColor": "#E0E0E0",
          "width": "73px",
        }
      }
      tabIndex="0"
    >
      <span
        className="text"
      >
        .
      </span>
    </div>
    <div
      className="button"
      onClick={[Function]}
      onKeyPress={[Function]}
      role="button"
      style={
        Object {
          "backgroundColor": "#F4924B",
          "width": "73px",
        }
      }
      tabIndex="0"
    >
      <span
        className="text"
      >
        =
      </span>
    </div>
  </div>
</div>
`;

exports[`testing operate and calculate function renders correctly for BarDisplay component 1`] = `
<div
  className="bar-display"
>
  <div
    className="writting"
    role="none"
  >
    12
  </div>
</div>
`;

exports[`testing operate and calculate function renders correctly for Button component 1`] = `
<div
  className="button"
  onClick={[Function]}
  onKeyPress={[Function]}
  role="button"
  style={
    Object {
      "backgroundColor": "blue",
      "width": "100px",
    }
  }
  tabIndex="0"
>
  <span
    className="text"
  >
    1
  </span>
</div>
`;

exports[`testing operate and calculate function renders correctly for Home component 1`] = `
<div
  style={
    Object {
      "fontSize": 19,
      "margin": "auto",
      "marginTop": "5%",
      "textAlign": "center",
      "width": "75%",
    }
  }
>
  <h1>
    Welcome to our page
  </h1>
  <p
    style={
      Object {
        "textAlign": "start",
      }
    }
  >
    Lorem Ipsum is simply dummy text of the printing and typesetting industry. Lorem Ipsum has been the industry&lsquo s standard dummy text ever since the 1500s, when an unknown printer took a galley of type and scrambled it to make a type specimen book. It has survived not only five centuries, but also the leap into electronic typesetting, remaining essentially unchanged. It was popularised in the 1960s with the release of Letraset sheets containing Lorem Ipsum passages, and more recently with desktop publishing software like Aldus PageMaker including versions of Lorem Ipsum.
  </p>
  <p
    style={
      Object {
        "textAlign": "start",
      }
    }
  >
    It is a long established fact that a reader will be distracted by the readable content of a page when looking at its layout. The point of using Lorem Ipsum is that it has a more-or-less normal distribution of letters, as opposed to using Content here, content here, making it look like readable English. Many desktop publishing packages and web page editors now use Lorem Ipsum as their default model text, and a search for lorem ipsum will uncover many web sites still in their infancy. Various versions have evolved over the years, sometimes by accident, sometimes on purpose (injected humour and the like).
  </p>
</div>
`;

exports[`testing operate and calculate function renders correctly for NavBar component 1`] = `
<div
  className="navbar"
>
  <div
    className="title"
  >
    <h1>
      Math Magicians
    </h1>
  </div>
  <div
    className="link"
  >
    <ul>
      <li>
        <a
          aria-current="page"
          className="active"
          href="/"
          onClick={[Function]}
          style={Object {}}
        >
          Home
        </a>
      </li>
      <li>
        <a
          aria-current={null}
          href="/calc"
          onClick={[Function]}
        >
          Calculator
        </a>
      </li>
      <li>
        <a
          aria-current={null}
          href="/quote"
          onClick={[Function]}
        >
          Quote
        </a>
      </li>
    </ul>
  </div>
</div>
`;

exports[`testing operate and calculate function renders correctly for Quote component 1`] = `
<div
  style={
    Object {
      "fontSize": 19,
      "margin": "auto",
      "marginTop": "5%",
      "textAlign": "center",
      "width": "70%",
    }
  }
>
  <p>
    Mathematics is about numbers, equations, computations
    <br />
    or algorithms: it is about understanding. -William Paul Thurston
  </p>
</div>
`;
